example/plugins/c-api/txn_data_sink/txn_data_sink.cc

```diff
@@ -117,43 +117,122 @@ client_reader(TSCont contp, TSEvent event, void *edata)
   return 0;
 }
 
+/** The handler for transaction data sink for response bodies. */
 int
-enable_agent_check(TSHttpTxn txnp)
+response_body_reader(TSCont contp, TSEvent event, void *edata)
 {
-  TSMBuffer req_buf;
-  TSMLoc req_loc;
-  int zret = 0;
+  SinkData *data = static_cast<SinkData *>(TSContDataGet(contp));
+
+  // If we got closed, we're done.
+  if (TSVConnClosedGet(contp)) {
+    delete data;
+    TSContDestroy(contp);
+    return 0;
+  }
+
+  TSVIO input_vio = TSVConnWriteVIOGet(contp);
 
+  if (data == nullptr) {
+    data = new SinkData;
+    TSContDataSet(contp, data);
+  }
+  return body_reader_helper(event, input_vio, data->response_body_bytes);
+}
+
+/** The handler for transaction data sink for request bodies. */
+int
+request_body_reader(TSCont contp, TSEvent event, void *edata)
+{
+  SinkData *data = static_cast<SinkData *>(TSContDataGet(contp));
+
+  // If we got closed, we're done.
+  if (TSVConnClosedGet(contp)) {
+    delete data;
+    TSContDestroy(contp);
+    return 0;
+  }
+
+  TSVIO input_vio = TSVConnWriteVIOGet(contp);
+
+  if (data == nullptr) {
+    data = new SinkData;
+    TSContDataSet(contp, data);
+  }
+  return body_reader_helper(event, input_vio, data->request_body_bytes);
+}
+
+/** A helper function for common logic between request_sink_requested and
+ * response_sink_requested. */
+bool
+sink_requested_helper(TSHttpTxn txnp, std::string_view header)
+{
+  bool field_found = false;
```

Comment:
Don't bother with that. Declare `field` here and then `return field != nullptr;`.